lib/python/qmk/cli/c2json.py

```diff
@@ -0,0 +1,324 @@
+
+    Parameters:
+    path -- path to the keymap.h we are resolving for
+
+    Returns:
+    A Path object for the keyboard.h found or None if we can't find one
+
+    Signature:
+    (path: Path) -> Path
+    """
+
+    def header_name(directory: Path) -> str:
+        return directory.name + '.h'
+
+    # path will look like
+    # keyboards/NAME/SUBNAME/keymaps/keymap.c
+    keyboard_h = None
+    for part in iter_upwards(path):
+        # if we hit a root level subdirectory we've overshot it
+        if part.name in ('keyboards', 'users', 'layouts'):
+            return keyboard_h
+
+        header = part / header_name(part)
+        if header.exists():
+            keyboard_h = header
+
+
+def get_qmk_root(from_path):
+    """
+    Finds the path of the QMK repository root relative to a keymaps directory
+
+    Parameters:
+    from_path -- Path to find the QMK root relative to
+
+    Signature:
+    (from_path: Path) -> Path
+    """
+    for part in iter_upwards(from_path.absolute()):
+        if (part / 'keyboards').exists():
+            return part
+
+
+def recursive_subdirs(from_path):
+    """
+    Generator that walks down the filesystem tree and gets all subdirectories
+    under the Path ``from_path``
+
+    Parameters:
+    from_path -- Path to walk from
+    """
+    yield from_path
+    for dirpath, dirs, files in os.walk(from_path):
+        yield from (Path(dirpath) / d for d in dirs)
+
+
+def parse_keymap_c(path):
+    """
+    Loads a C AST from the given keymap.c file.
+
+    Parameters:
+    path -- Path of the keymap.c to parse
+    """
+    qmk_root = get_qmk_root(path)
+    keyboard_h = find_topmost_keyboard_h(path)
+
+    # I'm guessing make probably doesn't just put the entire source tree as
+    # an include path??
+    def subdir_includes(parent: Path):
+        return ('-I{}'.format(parent / d) for d in recursive_subdirs(parent))
+
+    # we need to use the GNU C parser so attributes in system headers and
+    # other inconvenient things don't cause our parse to fail
+    return pycparser.parse_file(
+        str(path),
+        parser=GnuCParser(),
+        use_cpp=True,
+        cpp_args=[
+            '-DQMK_KEYBOARD_H="{}"'.format(keyboard_h),
+            '-std=gnu99',
+            *subdir_includes(keyboard_h.parent),
+            *subdir_includes(qmk_root / 'quantum'),
+            *subdir_includes(qmk_root / 'tmk_core'),
+        ]
+    )
+
+
+def find_keymaps_in_ast(ast):
+    """
+    Finds the keymaps object inside the full C AST of the keymap.c file
+    """
+    for node in ast.ext:
+        if type(node) != Decl:
+            continue
+        # we need to check if it is 'extern' to avoid parsing the extern
+        # declaration of keymaps within the QMK source itself
+        if node.name == 'keymaps' and 'extern' not in node.storage:
+            return node
+    return None
+
+
+Key = str
+
+class Layer(NamedTuple):
+    name: str
+    keys: List[List[Key]]
```

Comment:
Using Key here seems to obfuscate rather than improve readability.

```suggestion
    keys: List[List[str]]
```